### ISSUE TITLE
Bug in get_information_matrix_from_point_clouds

### ISSUE THREAD

**Issue by jingxa** [IntelVCL/Open3D#347]
Hello ,sir!
i want to use pose graph with pcl;
but i have some problems;

---
When i learn codes of `GetInformationMatrixFromPointClouds`:
- first, i think that the function `GetRegistrationResultAndCorrespondences` finds some corresponding points in source and target clouds, and compute overlap `fitness` and `inlier_rmse`;
but ,like below:
https://github.com/IntelVCL/Open3D/blob/352d2f4ca0dd577b37053e946eec5f2cf5afc423/src/Core/Registration/Registration.cpp#L353-L355
target_kdtree finds corresponding points!

 But i think "target_kdtree" should input transformed source points!

https://github.com/IntelVCL/Open3D/blob/352d2f4ca0dd577b37053e946eec5f2cf5afc423/src/Core/Registration/Registration.cpp#L62-L71
target_kdtree directly search "original" source point 's neighbor in target clouds, i can't understand it;

---
 can u give me some idea? thanks!

**Comment by syncle** [comment]
Hi @jingxa. 

> i want to use pose graph with pcl;

You mean Open3D, right? :)

> But i think "target_kdtree" should input transformed source points!

I seriously reviewed the code. You're right it is not in the right shape as you mentioned. It should be fixed. Let me begin some experiment by putting transformed point cloud into ``GetInformationMatrixFromPointClouds``, and will make a new pull request.

**Comment by jingxa** [comment]
@syncle Ha ,thanks!

---
My linear algebra is bad ;

---
Secoond, I don't understand some Equation in [Registration about Robust Reconstruction of Indoor Scenes](http://redwood-data.org/indoor/registration.html)

In equation (6)-(12)：
I think equation (7) was minimized to obtain  feasible : ![](https://latex.codecogs.com/gif.latex?%24T_%7Bji%7D%24)
but in equation (7):
![](https://latex.codecogs.com/gif.latex?%24%20%5Cleft%20%5C%7C%20w%5Ctimes%20q%5E%7B*%7D&plus;%20t%20%5Cright%20%5C%7C%5E%7B2%7D%24)
how does it become equation (8)?
![](https://latex.codecogs.com/gif.latex?%24%20%5Cleft%20%5C%7C%20%5B-%5Bq%5E%7B*%7D%5D_%7Bx%7D%20%5Cleft%20%7C%20I%5D%5Cxi%20%5D%20%5Cright%20%5C%7C%5E%7B2%7D%24)

---
![](https://latex.codecogs.com/gif.latex?-%5Bq%5E%7B*%7D%5D_%7Bx%7D):  the skew-symmetric matrix; what it's shape?
I think q* 's shape is (4,1),then skew_symmetric should be (4,4);
But ![](https://latex.codecogs.com/gif.latex?%5Cxi) 's shpe is (6,1),
so,![](https://latex.codecogs.com/gif.latex?%5B-%5Bq%5E%7B*%7D%5D_%7Bx%7D%20%5Cleft%7C%20I%20%5D) should be (n,6);

---
finally, i can't understand equation(8) and later how to work! 

can you give me some idea? thanks ！

**Comment by syncle** [comment]
> equation (7) to equation (8)

\xi contains [w,t]. By definition of vector cross product, it is equivalent to multiply vector with skew-symmetric matrix. Please check Cross product section in https://en.wikipedia.org/wiki/Skew-symmetric_matrix.

> what is the shape?

skew symmetric matrix is 3x3, and I is 3x3 so [-[q^*]_x|I] is 3x6 matrix.

>  i can't understand equation(8) and later how to work!

G is just redefinition of [-[q^*]_x|I], and the later equation is just re-writing quadratic equation. The key idea is to make information matrix \Lambda separable. As \Lambda can be pre-computed, the optimization does not require to find correspondences in every iteration. This is the core idea for the fast optimization.

**Comment by jingxa** [comment]
@syncle Oh ,very thanks !

---
Sorry for my bad math; I have another problem;
- <a href="https://www.codecogs.com/eqnedit.php?latex=T_{ji}^{*}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?T_{ji}^{*}" title="T_{ji}^{*}" /></a>:  ground truth for q to p;

- <a href="https://www.codecogs.com/eqnedit.php?latex=T_{ji}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?T_{ji}" title="T_{ji}" /></a> : sufficiently close for q to p;

then, we can make (2) close to (3): 
first,the equation(2)
<a href="https://www.codecogs.com/eqnedit.php?latex=\left&space;\|&space;p^{*}-T_{ji}q^{*}&space;\right&space;\|^{2}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\left&space;\|&space;p^{*}-T_{ji}q^{*}&space;\right&space;\|^{2}" title="\left \| p^{*}-T_{ji}q^{*} \right \|^{2}" /></a>
to the equation(3):
<a href="https://www.codecogs.com/eqnedit.php?latex=\left&space;\|&space;T_{ji}^{*}q^{*}&space;-&space;T_{ji}q^{*}&space;\right&space;\|^{2}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\left&space;\|&space;T_{ji}^{*}q^{*}&space;-&space;T_{ji}q^{*}&space;\right&space;\|^{2}" title="\left \| T_{ji}^{*}q^{*} - T_{ji}q^{*} \right \|^{2}" /></a>

I don't know wether q* is in homogeneous coordinates or European coordinates,but it is not necessary;
finally ,to the equation(12):
<a href="https://www.codecogs.com/eqnedit.php?latex=\xi&space;^{T}(\sum&space;G_{q^{*}}^{T}G_{q^{*}})\xi" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\xi&space;^{T}(\sum&space;G_{q^{*}}^{T}G_{q^{*}})\xi" title="\xi ^{T}(\sum G_{q^{*}}^{T}G_{q^{*}})\xi" /></a>

---
<a href="https://www.codecogs.com/eqnedit.php?latex=\Lambda&space;=&space;\sum_{p,p\epsilon&space;k_{ij}&space;}G_{p}^{T}G_{p}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\Lambda&space;=&space;\sum_{p,p\epsilon&space;k_{ij}&space;}G_{p}^{T}G_{p}" title="\Lambda = \sum_{p,p\epsilon k_{ij} }G_{p}^{T}G_{p}" /></a>
is constant, so we can pre-compute it;



---
- 1. In your article , you pre-compute 👍 



How to optimize the last equation ?
<a href="https://www.codecogs.com/eqnedit.php?latex=\xi^{T}\Lambda&space;\xi" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\xi^{T}\Lambda&space;\xi" title="\xi^{T}\Lambda \xi" /></a>

**Comment by syncle** [comment]
> by the way, is it wrong

I guess that might be not correct, but I am not author of that webpage. @qianyizh, can you check?

> I don't know wether q* is in homogeneous coordinates

Yes. It is homogeneous coordinate. T is 4x4 matrix.

> How to optimize the last equation ?

It is conventional convex optimization problem. You can compute gradient for that, and update \xi using that gradient. Please check https://github.com/IntelVCL/Open3D/blob/master/src/Core/Registration/GlobalOptimization.cpp. We have LM optimization there :)

**Comment by qianyizh** [comment]
It is not wrong.
Tji* q* is approximately equal to p*
(since they are correspondence pair, and Tji* is the ground truth transformation)
replace p* with Tji* q*, we get (3) from (2).

**Comment by jingxa** [comment]
@syncle  @qianyizh  yes, thanks , i am late , i mean that Tji* is right, but "Tij" is wrong??

**Comment by qianyizh** [comment]
Yes you are right. That is a typo, :)
